### PR TITLE
[Clang] Implement CWG2813: Class member access with prvalues

### LINKED ISSUE #100314: CWG2813: extra move in calling of explicit object member function
This program
```
struct A {
    A() {}
    A(A&&) = delete;
    void f(this A) {}
};

int main() {
    A{}.f();
}
```
must be accepted per [CWG2813](https://cplusplus.github.io/CWG/issues/2813.html), and GCC does accept it, but not Clang unfortunately.

Online demo: https://gcc.godbolt.org/z/6x7PTsE4M

Related discussion: https://stackoverflow.com/q/78785176/7325599

### LINKED ISSUE #100341: [Clang] Deducing this does not work correctly with move-only types.
Move only types have to be explicitly cast to r-value for the move constructor to be used. https://godbolt.org/z/fabdhKKEj

```
#include <utility>

struct X {
    X() = default;
    X(X&&) = default;
    void operator()(this X);
};

void fail() {
    X()();
    [x = X{}](this auto) {}();
}
void pass() {
    std::move(X())();
    std::move([x = X{}](this auto) {})();
}
```
```
<source>:10:5: error: call to implicitly-deleted copy constructor of 'X'
   10 |     X()();
      |     ^~~
<source>:5:5: note: copy constructor is implicitly deleted because 'X' has a user-declared move constructor
    5 |     X(X&&) = default;
      |     ^
<source>:6:27: note: passing argument to parameter here
    6 |     void operator()(this X);
      |                           ^
<source>:11:5: error: call to implicitly-deleted copy constructor of '(lambda at <source>:11:5)'
   11 |     [x = X{}](this auto) {}();
      |     ^~~~~~~~~~~~~~~~~~~~~~~
<source>:11:6: note: copy constructor of '(lambda at <source>:11:5)' is implicitly deleted because field '' has a deleted copy constructor
   11 |     [x = X{}](this auto) {}();
      |      ^
<source>:5:5: note: copy constructor is implicitly deleted because 'X' has a user-declared move constructor
    5 |     X(X&&) = default;
      |     ^
<source>:11:24: note: passing argument to parameter here
   11 |     [x = X{}](this auto) {}();
      |                        ^
```

### PR DESCRIPTION
**Pull request by cor3ntin**
This is a rebase of #95112 with my own feedback apply as @MitalAshok has been inactive for a while.
It's fairly important this makes clang 20 as it is a blocker for #107451 

--- 

[CWG2813](https://cplusplus.github.io/CWG/issues/2813.html)

prvalue.member_fn(expression-list) now will not materialize a temporary for prvalue if member_fn is an explicit object member function, and prvalue will bind directly to the object parameter.

The E1 in E1.static_member is now a discarded-value expression, so if E1 was a call to a [[nodiscard]] function, there will now be a warning. This also affects C++98 with [[gnu::warn_unused_result]] functions.

This should not affect C where TemporaryMaterializationConversion is a no-op.

Closes #100314
Fixes #100341

### CODE REVIEW

File: clang/lib/Sema/SemaExprMember.cpp [hunk 1]
```diff
@@ -1162,6 +1195,9 @@ Sema::BuildMemberReferenceExpr(Expr *BaseExpr, QualType BaseExprType,
       valueKind = VK_PRValue;
       type = Context.BoundMemberTy;
```
Comment (Sirraide) [inline clang/lib/Sema/SemaExprMember.cpp:1196]: Don’t we need to call `MakeGLValue()` in the then-clause here too?

File: clang/lib/Sema/SemaStmt.cpp [hunk 2]
```diff
@@ -391,16 +397,39 @@ void Sema::DiagnoseUnusedExprResult(const Stmt *S, unsigned DiagID) {
...
       !E->getType()->isArrayType()) {
-    Diag(Loc, diag::warn_unused_volatile) << R1 << R2;
+    S.Diag(Loc, diag::warn_unused_volatile) << R1 << R2;
     return;
   }
 
+  // Don't diagnose discarded left of dot in static class member access
+  // because its type is "used" to determine the class to access
+  //if (DiagID == diag::warn_discarded_class_member_access)
+  //  return;
```
Comment (Sirraide) [inline clang/lib/Sema/SemaStmt.cpp]: I think this can be deleted at this point?

File: clang/docs/ReleaseNotes.rst [hunk 3]
```diff
@@ -660,7 +665,7 @@ Improvements to Clang's diagnostics
 
 - Clang now diagnoses dangling references for C++20's parenthesized aggregate initialization (#101957).
 
-- Fixed a bug where Clang would not emit ``-Wunused-private-field`` warnings when an unrelated class 
+- Fixed a bug where Clang would not emit ``-Wunused-private-field`` warnings when an unrelated class
```
Comment (erichkeane) [inline clang/docs/ReleaseNotes.rst:668]: couple of unrelated changes, here and 831 :)

File: clang/include/clang/Sema/Sema.h [hunk 4]
```diff
@@ -10659,6 +10659,11 @@ class Sema final : public SemaBase {
                            SourceLocation EndLoc);
   void ActOnForEachDeclStmt(DeclGroupPtrTy Decl);
 
+  /// DiagnoseDiscardedExprMarkedNodiscard - Given an expression that is
+  /// semantically a discarded-value expression, diagnose if any [[nodiscard]]
+  /// value has been discarded
```
Comment (erichkeane) [inline clang/include/clang/Sema/Sema.h:10664]:
```suggestion
  /// value has been discarded.
```

File: clang/lib/Sema/SemaExprMember.cpp [hunk 5]
```diff
@@ -1128,26 +1119,68 @@ Sema::BuildMemberReferenceExpr(Expr *BaseExpr, QualType BaseExprType,
     BaseExpr = BuildCXXThisExpr(Loc, BaseExprType, /*IsImplicit=*/true);
   }
 
+  // C++17 [expr.ref]p2, per CWG2813:
+  //   For the first option (dot), if the id-expression names a static member or
+  //   an enumerator, the first expression is a discarded-value expression; if
+  //   the id-expression names a non-static data member, the first expression
+  //   shall be a glvalue.
+  auto MakeDiscardedValue = [&] {
```
Comment (erichkeane) [inline clang/lib/Sema/SemaExprMember.cpp:1127]: I would prefer some more descriptive names here?  Perhaps `ConvertBaseExprToDiscardedValue` and `ConvertBaseExprToGLValue` ?

**Review by erichkeane — APPROVED** [review]
A few nits, else LGTM.

**Comment by shafik** [comment]
It looks like this is linked to: https://github.com/llvm/llvm-project/issues/131410

and it is not clear to me the diagnostic in these cases makes a lot of sense or at least the motivation does not feel strong to me.

This new approach feels evolutionary to me and if this was the intent then I am not sure we should implement it.

**Pull request merged**